### PR TITLE
Clean up attention implementation

### LINKED ISSUE #59: torch.compile and check pytorch nightly
Suggestions from today's session with Horace - remove compile flag (we always want to torch.compile flex attention) and call torch.compile on model in train script. Optionally, check pytorch nightly to see if we can compile block mask creation in a better way

### LINKED ISSUE #70: Parameter names should be more explicit
The parameter names used for the various layers can be changed to be more explicit. In addition to having docstrings for the parameter functions, the names can be changed so that they hint more clearly at their purpose.

### LINKED ISSUE #72: Missing docstrings explaining input and output tensor shapes
It would be helpful to have explicit tensor shapes provided in the documentation for the layers so that users can easily tell how their tensors should be shaped before passing to each layer.

### LINKED ISSUE #76: Clean up Attention structures in TMRC
The Attention module in TMRC is a bit confusing with the flash and flex related options. It would clean up the structure by using some inheritance structure to define the underlying attention implementation. Then for each mask, we can define a new attention subclass inheriting the optimal implementation, so that users don't have to worry about what is optimal.

### PR DESCRIPTION
**Pull request by timothyngo**
Closes #76 
Closes #59
Closes #70 
Closes #72 

Rewrites the attention implementation to be more modular, so that it is easier to construct attention layers for different types of masks. This rewrite will also make it easier to understand and test the code.

Key points:
- A clear inheritance structure for self-attention layers has been defined
  - All self-attention layers must inherit from `SelfAttentionBase` or one of its direct children
  - The children of `SelfAttentionBase` implement a specific attention algorithm (e.g. Flash or Flex Attention)
  - The grandchildren of `SelfAttentionBase` implement a specific type of mask. The child it inherits from is chosen based on which algorithm is optimal for that mask type.
- Parameters have been cleaned up to have clearer names and easier to use when classes are imported
- Documentation has been added to describe all parameters

### CODE REVIEW

File: tests/test_gpt/test_config.yaml [hunk 1]
```diff
@@ -16,10 +16,7 @@ model:
   ln_bias: True
   cls_head_bias: True
   activation: relu
-  flash: True
-  flex: False
-  mask: causal_document
-  compile_flex: False
+  mask: causal
```
Comment (timothyngo) [inline tests/test_gpt/test_config.yaml:19]: This is equivalent to the previous code. The old code always used a causal mask when using FlashAttention, regardless of the provided mask.

**Comment by timothyngo** [comment]
This PR has a lot of changes, but most of it is documentation. The main code change is the refactoring of the attention layers. Most other changes were either renaming parameters or adding detailed documentation.

**Comment by nikhilanand91** [comment]
Looked over the changes -- all looks good to me.  Eventually, we might want to update our documentation to show how a new user can use these patterns to create a custom attention implementation (e.g., with a different mask), but that can be separate issue.

**Comment by timothyngo** [comment]
> Looked over the changes -- all looks good to me. Eventually, we might want to update our documentation to show how a new user can use these patterns to create a custom attention implementation (e.g., with a different mask), but that can be separate issue.

This is a good point! Will make an issue for this once this PR is merged.

File: src/tmrc/core/models/components/decoder.py [hunk 2]
```diff
@@ -1,198 +1,460 @@
...
-from torch.nn.attention.flex_attention import flex_attention
+from torch.nn.attention.flex_attention import BlockMask, create_block_mask
+
+T = TypeVar("T")
 
-from tmrc.core.models.components import ACTIVATION_REGISTRY
+if torch.cuda.is_available() and torch.cuda.device_count() > 0:
+    DEVICE = "cuda"
+else:
+    DEVICE = "cpu"
```
Comment (amazloumi) [inline src/tmrc/core/models/components/decoder.py:15]: This code here make duplication with what we have in `platform` class. not a blocker but if possible to pass as argument would be good.

**Review by amazloumi — COMMENTED** [review]
Then no need for mask registry and the code block:
```
"""Masks"""
@register_mask("causal")
def causal_mask(score, b, h, q_idx, kv_idx):
    return torch.where(q_idx >= kv_idx, score, -float("inf"))

@register_mask("causal_document")
def get_doc_causal_mask_fn(doc_mask_ids):
    def document_causal_mask(b, h, q_idx, kv_idx):
        causal_mask = q_idx >= kv_idx
        document_mask = doc_mask_ids[b, q_idx] == doc_mask_ids[b,kv_idx]
        return causal_mask & document_mask
    return document_causal_mask
```
better to remove it and make sure tmrc is working correctly with the new structure of attention block without the old version.

**Comment by timothyngo** [comment]
> Then no need for mask registry and the code block:
> 
> ```
> """Masks"""
> @register_mask("causal")
> def causal_mask(score, b, h, q_idx, kv_idx):
>     return torch.where(q_idx >= kv_idx, score, -float("inf"))
> 
> @register_mask("causal_document")
> def get_doc_causal_mask_fn(doc_mask_ids):
>     def document_causal_mask(b, h, q_idx, kv_idx):
>         causal_mask = q_idx >= kv_idx
>         document_mask = doc_mask_ids[b, q_idx] == doc_mask_ids[b,kv_idx]
>         return causal_mask & document_mask
>     return document_causal_mask
> ```
> 
> better to remove it and make sure tmrc is working correctly with the new structure of attention block without the old version.

I just removed these registries.

**Review by amazloumi — APPROVED** [review]
LGTM!

File: src/tmrc/core/models/components/decoder.py (hunk 2 above)
Comment (timothyngo) [reply src/tmrc/core/models/components/decoder.py:12]: I changed the code to use Platform internally and removed this global.

File: src/tmrc/core/models/components/decoder.py [hunk 3]
```diff
@@ -1,198 +1,458 @@
...
+            return causal_mask & document_mask
+
+        return create_block_mask(
+            document_causal_mask,
+            mask_data.shape[0],
+            None,
+            mask_data.shape[1],
+            mask_data.shape[1],
+            _compile=True,
+            device=self.platform.get_device_type(),
```
Comment (amazloumi) [inline src/tmrc/core/models/components/decoder.py:318]: `platform .get_device_type()` only returns `cuda` with no index, use `platform.get_device_str()` to return with index like `cuda:0` for GPU 0 and so on. later in multi-gpu setting each process knows about their gpu index and generate different device strings like "cuda:0", "cuda:1" and so on.

Comment (timothyngo) [reply src/tmrc/core/models/components/decoder.py]: Also, just changed the `get_device_type` call to `get_device_str`.

Comment (mbsabath) [reply src/tmrc/core/models/components/decoder.py]: Ah yeah, checked the `Platform` code and looks like we're already handling it, should work both with and without cgroups limiting visible GPUs

**Review by mbsabath — APPROVED** [review]
LGTM!
